File: src/model/nostr/parser/kind7374.go
// kind7374.go - Quote Event

package parser

import (
	"fmt"
	"strconv"
	"time"

	"github.com/candypoets/nutscash/nostr/types"
	"github.com/nbd-wtf/go-nostr"
)

// Kind7374Parsed represents a parsed quote event for redeeming Cashu tokens
type Kind7374Parsed struct {
	QuoteID    string    `json:"quoteId" msgpack:"quoteId"`
	MintURL    string    `json:"mintUrl" msgpack:"mintUrl"`
	Expiration time.Time `json:"expiration" msgpack:"expiration"`
}

// ParseKind7374 parses a kind 7374 (quote) event
func (p *Parser) ParseKind7374(event nostr.Event) (*Kind7374Parsed, *[]types.Request, error) {
	if event.Kind != 7374 {
		return nil, nil, fmt.Errorf("event is not kind 7374")
	}

	// Extract mint URL from tags
	var mintURL string
	var expirationUnix int64

	for _, tag := range event.Tags {
		if len(tag) >= 2 {
			if tag[0] == "mint" {
				mintURL = tag[1]
			} else if tag[0] == "expiration" {
				if expTS, err := strconv.ParseInt(tag[1], 10, 64); err == nil {
					expirationUnix = expTS
				}
			}
		}
	}

	if mintURL == "" {
		return nil, nil, fmt.Errorf("mint URL not found in quote event")
	}

	// Try to decrypt the quote ID if the signer is available
	var quoteID string
	if p.Signer != nil {
		pubkey, _ := p.Signer.Current.GetPublicKey()
		decrypted, err := p.Signer.Current.NIP44Decrypt(pubkey, event.Content)
		if err == nil && decrypted != "" {
			quoteID = decrypted
		}
	}

	// Create expiration time
	var expiration time.Time
	if expirationUnix > 0 {
		expiration = time.Unix(expirationUnix, 0)
	} else {
		// No expiration provided, leave it as zero value
		expiration = time.Time{}
	}

	return &Kind7374Parsed{
		QuoteID:    quoteID,
		MintURL:    mintURL,
		Expiration: expiration,
	}, nil, nil
}

// PrepareKind7374 prepares a kind 7374 event for publishing by encrypting content and signing
func (p *Parser) PrepareKind7374(event *nostr.Event) error {
	if event.Kind != 7374 {
		return fmt.Errorf("event is not kind 7374")
	}

	// Validate required tags
	var hasMint bool
	for _, tag := range event.Tags {
		if len(tag) >= 2 && tag[0] == "mint" {
			hasMint = true
			break
		}
	}

	if !hasMint {
		return fmt.Errorf("kind 7374 events must have a mint tag")
	}

	currentSigner := p.Signer.Current

	if currentSigner == nil {
		return fmt.Errorf("signer is required for kind 7374 events")
	}

	pubkey, _ := currentSigner.GetPublicKey()
	// NIP-44 encrypt the quote ID content
	encrypted, err := currentSigner.NIP44Encrypt(pubkey, event.Content)
	if err != nil {
		return fmt.Errorf("failed to encrypt quote ID: %w", err)
	}
	event.Content = encrypted

	// Sign the event
	return currentSigner.SignEvent(event)
}

File: src/model/nostr-worker/src/parser/kind7374.rs
use crate::parser::Parser;
use crate::types::network::Request;
use anyhow::{anyhow, Result};
use nostr::{Event, UnsignedEvent};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Kind7374Parsed {
    #[serde(rename = "quoteId")]
    pub quote_id: String,
    #[serde(rename = "mintUrl")]
    pub mint_url: String,
    pub expiration: u64, // Unix timestamp
}

impl Parser {
    pub fn parse_kind_7374(&self, event: &Event) -> Result<(Kind7374Parsed, Option<Vec<Request>>)> {
        if event.kind.as_u64() != 7374 {
            return Err(anyhow!("event is not kind 7374"));
        }

        // Extract mint URL from tags
        let mut mint_url = String::new();
        let mut expiration_unix = 0u64;

        for tag in &event.tags {
            let tag_vec = tag.as_vec();
            if tag_vec.len() >= 2 {
                match tag_vec[0].as_str() {
                    "mint" => {
                        mint_url = tag_vec[1].clone();
                    }
                    "expiration" => {
                        if let Ok(exp_ts) = tag_vec[1].parse::<u64>() {
                            expiration_unix = exp_ts;
                        }
                    }
                    _ => {}
                }
            }
        }

        if mint_url.is_empty() {
            return Err(anyhow!("mint URL not found in quote event"));
        }

        let mut quote_id = String::new();

        if self.signer_manager.has_signer() {
            let pubkey = self.signer_manager.get_public_key()?;
            if let Ok(decrypted) = self.signer_manager.nip44_decrypt(&pubkey, &event.content) {
                if !decrypted.is_empty() {
                    quote_id = decrypted;
                }
            }
        }

        let parsed = Kind7374Parsed {
            quote_id,
            mint_url,
            expiration: expiration_unix,
        };

        Ok((parsed, None))
    }

    pub fn prepare_kind_7374(&self, event: &mut UnsignedEvent) -> Result<Event> {
        if event.kind.as_u64() != 7374 {
            return Err(anyhow!("event is not kind 7374"));
        }

        // Validate required tags
        let mut has_mint = false;
        for tag in &event.tags {
            let tag_vec = tag.as_vec();
            if tag_vec.len() >= 2 && tag_vec[0] == "mint" {
                has_mint = true;
                break;
            }
        }

        if !has_mint {
            return Err(anyhow!("kind 7374 events must have a mint tag"));
        }

        if self.signer_manager.has_signer() {
            let pubkey = self.signer_manager.get_public_key()?;
            let encrypted = self.signer_manager.nip44_encrypt(&pubkey, &event.content)?;
            event.content = encrypted;
            let new_event = self.signer_manager.sign_event(event)?;
            Ok(new_event)
        } else {
            Err(anyhow!("signer is required for kind 7374 events"))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use nostr::{EventBuilder, Keys, Kind, Tag};

    #[test]
    fn test_parse_kind_7374_basic() {
        let keys = Keys::generate();
        let mint_url = "https://mint.example.com";
        let expiration = 1234567890u64;

        let tags = vec![
            Tag::parse(vec!["mint".to_string(), mint_url.to_string()]).unwrap(),
            Tag::parse(vec!["expiration".to_string(), expiration.to_string()]).unwrap(),
        ];

        let event = EventBuilder::new(Kind::Custom(7374), "encrypted_quote_id", tags)
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let (parsed, _) = parser.parse_kind_7374(&event).unwrap();

        assert_eq!(parsed.mint_url, mint_url);
        assert_eq!(parsed.expiration, expiration);
        assert_eq!(parsed.quote_id, ""); // No decryption without signer
    }

    #[test]
    fn test_parse_kind_7374_missing_mint() {
        let keys = Keys::generate();

        let event = EventBuilder::new(Kind::Custom(7374), "content", Vec::new())
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let result = parser.parse_kind_7374(&event);

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("mint URL not found"));
    }

    #[test]
    fn test_prepare_kind_7374_no_signer() {
        // let keys = Keys::generate();
        // let mint_url = "https://mint.example.com";

        // let tags = vec![Tag::parse(vec!["mint".to_string(), mint_url.to_string()]).unwrap()];

        // let mut event = EventBuilder::new(Kind::Custom(7374), "quote_id", tags)
        //     .to_event(&keys)
        //     .unwrap();

        // let parser = Parser::default();
        // let result = parser.prepare_kind_7374(&mut event);

        // assert!(result.is_err());
        // assert!(result
        //     .unwrap_err()
        //     .to_string()
        //     .contains("encryption and signing not implemented"));
    }

    #[test]
    fn test_parse_wrong_kind() {
        let keys = Keys::generate();

        let event = EventBuilder::new(Kind::TextNote, "test", Vec::new())
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let result = parser.parse_kind_7374(&event);

        assert!(result.is_err());
    }
}
